developer-docs/architecture/type-system.md

```diff
@@ -0,0 +1,260 @@
+
+In the case of `Asset`s, the literal data is a textual string, and the referenced file
+is an opaque blob.
+
+In the case of `Archive`s, the literal data is a map from strings to `Asset`s or `Archive`s,
+and the referenced file is a TAR archive, gzipped TAR archive, or ZIP archive.
+
+Each `Asset` or `Archive` also carries the SHA-256 hash of its contents. This hash can be
+used to uniquely identify the asset (e.g. for locally caching `Asset` or `Archive`
+contents).
+
+### Resource References
+
+A `ResourceReference` represents a reference to a resource. Although
+all that is necessary to uniquely identify a resource within the context of a stack is its
+URN, a `ResourceReference` also carries the resource's ID (if it is not a component) and
+the version of the provider that manages the resource. If the contents of the referenced
+resource must be inspected, the reference must be resolved by invoking the `getResource`
+function of the engine's builtin provider. Note that this is only possible if there is a 
+connection to the engine's resource monitor, e.g. within the scope of a call to `Construct`.
+This implies that resource references may not be resolved within calls to other 
+provider methods. Therefore, configuration values, custom resources and provider functions
+should not rely on the ability to resolve resource references, and should instead treat
+resource references  as either their ID (if present) or URN. If the ID is present and
+empty, it should be treated as an [`Unknown`](#unknowns).
+
+## Object Types
+
+Object types are defined as mapping from property names to property types. Duplicate
+property names are not allowed, and each property name maps to a single type.
+
+## `Promise<T>`
+
+A value of type `Promise<T>` represents the result of an asynchronous computation. Note
+that although computations may fail, failures cannot be handled at runtime, and cause a
+hard stop when attempting to access a `Promise<T>`'s concrete value.
+
+## `Output<T>`
+
+Perhaps the most important type in the Pulumi type system is `Output<T>`. A value of type
+`Output<T>` represents a node in a Pulumi program graph, and behaves like a `Promise<T>`
+that carries additional metadata that describes the resources on which the value depends,
+whether the value is known or unknown, and whether or not the value is secret. Readers
+familiar with functional programming will recognize `Output<T>` as a monad.
+
+### Dependencies
+
+If an `Output<T>` value is the result of a resource operation--e.g. if it is an output
+property of some resource--it is said to _depend_ on that resource.
+
+If a value of type `Output<T>` depends on a resource `R`, the result of any computation
+that depends on its concrete value also depends on `R`.
+
+### Unknowns 
+
+An `Output<T>` may be unknown if it depends on the result of a resource operation that
+will not be run because it is part of a `pulumi preview`. Previews typically produce for
```

Comment:
```suggestion
will not be run because it is part of a `pulumi preview`. Previews typically produce unknown for
```

Pretty sure a word was omitted here

Reply:
Fixed